geode-protobuf/src/main/java/org/apache/geode/protocol/protobuf/operations/GetAllRequestOperationHandler.java

```diff
@@ -50,26 +53,52 @@
-      }
-      return Success.of(RegionAPI.GetAllResponse.newBuilder().addAllEntries(entries).build());
-    } catch (UnsupportedEncodingTypeException ex) {
-      return Failure.of(ProtobufResponseUtilities.makeErrorResponse(
-          ProtocolErrorCode.VALUE_ENCODING_ERROR.codeValue, "Encoding not supported."));
-    } catch (CodecNotRegisteredForTypeException ex) {
-      return Failure.of(ProtobufResponseUtilities.makeErrorResponse(
-          ProtocolErrorCode.VALUE_ENCODING_ERROR.codeValue,
-          "Codec error in protobuf deserialization."));
+    Map<Boolean, List<Object>> resultsCollection = request.getKeyList().stream()
+        .map((key) -> processOneMessage(serializationService, region, key))
+        .collect(Collectors.partitioningBy(x -> x instanceof BasicTypes.Entry));
+    RegionAPI.GetAllResponse.Builder responseBuilder = RegionAPI.GetAllResponse.newBuilder();
+
+    for (Object entry : resultsCollection.get(true)) {
+      responseBuilder.addEntries((BasicTypes.Entry) entry);
+    }
+
+    for (Object entry : resultsCollection.get(false)) {
+      responseBuilder.addFailures((BasicTypes.KeyedError) entry);
     }
+
+    return Success.of(responseBuilder.build());
   }
 
+  private Object processOneMessage(SerializationService serializationService, Region region,
+      BasicTypes.EncodedValue key) {
+    try {
+      Object decodedKey = ProtobufUtilities.decodeValue(serializationService, key);
+      Object value = region.get(decodedKey);
+      return ProtobufUtilities.createEntry(serializationService, decodedKey, value);
+    } catch (CodecNotRegisteredForTypeException | UnsupportedEncodingTypeException ex) {
+      return BasicTypes.KeyedError.newBuilder().setKey(key)
+          .setError(BasicTypes.Error.newBuilder()
+              .setErrorCode(ProtocolErrorCode.VALUE_ENCODING_ERROR.codeValue)
+              .setMessage("Encoding not supported."))
+          .build();
+    } catch (org.apache.geode.distributed.LeaseExpiredException | TimeoutException e) {
+      return BasicTypes.KeyedError.newBuilder().setKey(key)
+          .setError(BasicTypes.Error.newBuilder()
+              .setErrorCode(ProtocolErrorCode.OPERATION_TIMEOUT.codeValue)
+              .setMessage("Operation timed out: " + e.getMessage()))
+          .build();
+    } catch (CacheLoaderException | PartitionedRegionStorageException e) {
+      return BasicTypes.KeyedError.newBuilder().setKey(key)
+          .setError(BasicTypes.Error.newBuilder()
+              .setErrorCode(ProtocolErrorCode.DATA_UNREACHABLE.codeValue)
+              .setMessage("Data unreachable: " + e.getMessage()))
+          .build();
+    } catch (NullPointerException | IllegalArgumentException e) {
+      return BasicTypes.KeyedError.newBuilder().setKey(key)
+          .setError(BasicTypes.Error.newBuilder()
+              .setErrorCode(ProtocolErrorCode.CONSTRAINT_VIOLATION.codeValue)
+              .setMessage("Invalid input: " + e.getMessage()))
+          .build();
```

Comment:
Do we have any tests that prove that these exceptions are thrown and correctly handled?

Reply:
👍 I think we should do this as part of a bigger refactor commit. Most of the message handlers build manually rather than using `ProtobufResponseUtilities.createAndLogErrorResponse`.